drivers/lora/sx1276.c

```diff
@@ -0,0 +1,717 @@
+		.count = 2,
+	};
+
+	if (!write) {
+		const struct spi_buf_set rx = {
+			.buffers = buf,
+			.count = 2
+		};
+
+		return spi_transceive(dev_data->spi, &dev_data->spi_cfg,
+				      &tx, &rx);
+	}
+
+	return spi_write(dev_data->spi, &dev_data->spi_cfg, &tx);
+}
+
+int sx1276_read(struct device *dev, u8_t reg_addr, u8_t *data, u8_t len)
+{
+	return sx1276_transceive(dev, reg_addr, false, data, len);
+}
+
+int sx1276_write(struct device *dev, u8_t reg_addr, u8_t byte)
+{
+	return sx1276_transceive(dev, reg_addr | BIT(7), true, &byte, 1);
+}
+
+int sx1276_fifo_write(struct device *dev, u8_t *data, u32_t data_len)
+{
+	int ret, i;
+
+	for (i = 0; i < data_len; i++) {
+		ret = sx1276_write(dev, SX1276_REG_FIFO, data[i]);
+		if (ret < 0) {
+			LOG_ERR("Unable to write FIFO");
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+int sx1276_fifo_read(struct device *dev, u8_t *data, u32_t data_len)
+{
+	int ret, i;
+	u8_t regval;
+
+	for (i = 0; i < data_len; i++) {
+		ret = sx1276_read(dev, SX1276_REG_FIFO, &regval, 1);
+		if (ret < 0) {
+			LOG_ERR("Unable to read FIFO");
+			return -EIO;
+		}
+
+		data[i] = regval;
+	}
+
+	return 0;
+}
+
+static int sx1276_lora_recv(struct device *dev, u8_t *data)
+{
+	struct sx1276_data *dev_data = dev->driver_data;
+	int ret;
+	u8_t regval, len, loc = 0;
+
+	/* Clear all flags */
+	ret = sx1276_write(dev, SX1276_REG_IRQ_FLAGS, 0xFF);
+	if (ret < 0) {
+		LOG_ERR("Unable to write IRQ_FLAGS");
+		return -EIO;
+	}
+
+	/* Set DIO0 to Rx done */
+	ret = sx1276_read(dev, SX1276_REG_DIO_MAPPING1, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read DIO_MAPPING1");
+		return -EIO;
+	}
+
+	regval &= ~REG_DIO_MAPPING1_DIO0_MASK;
+	ret = sx1276_write(dev, SX1276_REG_DIO_MAPPING1, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write DIO_MAPPING1");
+		return -EIO;
+	}
+
+	ret = sx1276_read(dev, SX1276_REG_OPMODE, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	/* FIFO cannot be accessed in sleep mode */
+	if (regval & SX1276_OPMODE_MODE_SLEEP) {
+		regval &= ~SX1276_OPMODE_MODE_MASK;
+		regval |= SX1276_OPMODE_MODE_STDBY;
+	}
+
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	/* Make use of full FIFO for Rx */
+	ret = sx1276_write(dev, SX1276_REG_FIFO_RX_BASE_ADDR, 0x00);
+	if (ret < 0) {
+		LOG_ERR("Unable to write FIFO Rx base addr");
+		return -EIO;
+	}
+
+	ret = sx1276_write(dev, SX1276_REG_FIFO_ADDR_PTR, 0x00);
+	if (ret < 0) {
+		LOG_ERR("Unable to write FIFO addr pointer");
+		return -EIO;
+	}
+
+	ret = sx1276_read(dev, SX1276_REG_OPMODE, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	regval &= ~SX1276_OPMODE_MODE_MASK;
+	regval |= SX1276_OPMODE_MODE_RXCONTINUOUS;
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	k_sem_take(&dev_data->data_sem, K_FOREVER);
```

Comment:
Wouldn't it be a nice to have if you pass the timeout parameter as a function argument?
fro example:
```C
static int sx1276_lora_recv(struct device *dev, u8_t *data, s32_t timeout)
```